src/ophyd_async/core/async_status.py

```diff
@@ -94,3 +86,75 @@ def __repr__(self) -> str:
         return f"<{type(self).__name__}, task: {self.task.get_coro()}, {status}>"
 
     __str__ = __repr__
+
+
+class AsyncStatus(AsyncStatusBase):
+    @classmethod
+    def wrap(cls: Type[AS], f: Callable[P, Awaitable]) -> Callable[P, AS]:
+        @functools.wraps(f)
+        def wrap_f(*args: P.args, **kwargs: P.kwargs) -> AS:
+            return cls(f(*args, **kwargs))
+
+        # type is actually functools._Wrapped[P, Awaitable, P, AS]
+        # but functools._Wrapped is not necessarily available
+        return cast(Callable[P, AS], wrap_f)
+
+
+class WatchableAsyncStatus(AsyncStatusBase, Generic[T]):
+    """Convert AsyncIterator of WatcherUpdates to bluesky Status interface."""
+
+    def __init__(
+        self,
+        iterator: AsyncIterator[WatcherUpdate[T]],
+        timeout_s: float = 0.0,
+    ):
+        self._watchers: list[Watcher] = []
+        self._start = time.monotonic()
+        self._timeout = self._start + timeout_s if timeout_s else None
+        self._last_update: WatcherUpdate[T] | None = None
+        super().__init__(self._notify_watchers_from(iterator))
+
+    async def _notify_watchers_from(self, iterator: AsyncIterator[WatcherUpdate[T]]):
+        async for update in iterator:
+            if self._timeout and time.monotonic() > self._timeout:
+                raise TimeoutError()
+            self._last_update = (
+                update
+                if update.time_elapsed is None
+                else replace(update, time_elapsed=time.monotonic() - self._start)
+            )
+            for watcher in self._watchers:
+                self._update_watcher(watcher, self._last_update)
+
+    def _update_watcher(self, watcher: Watcher, update: WatcherUpdate[T]):
+        vals = asdict(
+            update, dict_factory=lambda d: {k: v for k, v in d if v is not None}
+        )
+        watcher(**vals)
+
+    def watch(self, watcher: Watcher):
+        self._watchers.append(watcher)
+        if self._last_update:
+            self._update_watcher(watcher, self._last_update)
+
+    @classmethod
+    def wrap(
+        cls: Type[WAS],
+        f: Callable[P, AsyncIterator[WatcherUpdate[T]]],
+        timeout_s: float = 0.0,
+    ) -> Callable[P, WAS]:
+        """Wrap an AsyncIterator in a WatchableAsyncStatus. If it takes
+        'timeout_s' as an argument, this must be a float and it will be propagated
+        to the status."""
+
+        @functools.wraps(f)
+        def wrap_f(*args: P.args, **kwargs: P.kwargs) -> WAS:
```

Comment:
Does it work if we
```suggestion
        def wrap_f(*args: P.args, timeout: Optional[float] = None, **kwargs: P.kwargs) -> WAS:
```

Reply:
I couldn't get this to work - besides `ParamSpec` you can only have positional arguments at the beginning, which kind of defeats the purpose of having the paramspec at all if we don't want to enforce timeout being the first argument for all wrappable methods

Reply:
![Screenshot from 2024-04-18 16-29-50](https://github.com/bluesky/ophyd-async/assets/115003895/1ed21550-89a2-49a8-94ab-c08a945f580a)

Reply:
Having read [the spec](https://peps.python.org/pep-0612/#concatenating-keyword-parameters) I now understand the reasoning, so this makes sense